content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,608 @@
+so we created the `pad_collate_fn` function to process the necessary work within the `DataLoader`. Finally, 
+we will load the data using the vectorizer (within the `DataLoader` using the `pad_collate` function). This means that the addresses will be represented by word embeddings. 
+Also, the address components will be converted into categorical value (from 0 to 7).
+
+
+## The Training
+
+Now that we have all the components for the network, let's define our optimizer (Stochastic Gradient Descent) 
+([SGD](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)).
+
+```python
+optimizer = optim.SGD(full_network.parameters(), lr)
+```
+
+### Poutyne Experiment
+> Disclaimer: David is a dev on Poutyne, so we will present code using this framework. See the project [here](https://poutyne.org/).
+
+Let's create our experiment using Poutyne for automatically logging in the project root directory (`./`). We will also set
+the loss function and a batch metric (accuracy) to monitor the training.
+
+```python
+exp = Experiment("./", full_network, device=cuda_device, optimizer=optimizer,
+                 loss_function=cross_entropy, batch_metrics=["acc"])
+```
+
+Using our experiment, we can now launch the training as simply as
+
+```python
+exp.train(train_loader, valid_generator=valid_loader, epochs=epoch_number)
+```
+It will take around 6 minutes per epochs, so around an hour for the complete training.
+
+### Results
+The next figure shows the loss and the accuracy during our training (blue) and during our validation (orange) steps.
+After 10 epochs, we obtain a validation loss and accuracy of `0.01981` and `99.54701` respectively, which is pretty
+good for a first model. Also, we can see that our model did not seem to have overfitted.
+
+![loss_acc](graph/training_graph.png)
+
+## Bigger model
+
+It seems that our model performed pretty well, but just for fun, let's unleash the full potential of LSTM using a
+bidirectional approach (bidirectional LSTM). What it means is that instead of _simply_ viewing the sequence from the start to the end, we
+also train the model to see the sequence from the end to the start. It's important to state that the two directions are
+not shared, meaning that we _see_ the sequence in one direction at the time, but we gather the information from both directions into the 
+fully connected layer. That way, our model can get insight from both directions.
+
+Instead of using only one layer, let's use a bidirectional bi-LSTM, which means that we use two layers of hidden state for each direction.
+
+So, let's create the new LSTM and fully connected network.
+
+```python
+dimension = 300
+num_layer = 2
+bidirectional = True
+
+lstm_network = nn.LSTM(input_size=dimension,
+                       hidden_size=dimension,
+                       num_layers=num_layer,
+                       bidirectional=bidirectional,
+                       batch_first=True)
+
+input_dim = dimension * 2 #since bidirectional
```

Comment:
```suggestion
input_dim = dimension * (2 * int(bidirectional))  # since bidirectional
```

Reply:
The idea is better, but since we have a specific constructor for the bidirectional, I don't think this is necessary. In a unified one, it would be, but in this specific case, I don't think it's.